### PR TITLE
Parallelise the decoding of chunks coming from the cache.

### LINKED ISSUE #625: Parallelise chunk decoding
Part of #624

`Cache.FetchChunkData()` deserialises all chunks returned by memcached in a simple loop.  This could be parallelised when there are thousands to work through.

As [noted by jml](https://github.com/weaveworks/cortex/pull/603#discussion_r152310433) we already have a few implementations of "parallel map" - maybe we can re-use one here?

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #625

### CODE REVIEW

File: pkg/chunk/chunk_store_utils.go [hunk 1]
```diff
@@ -149,24 +201,41 @@ func ProcessCacheResponse(chunks []Chunk, keys []string, bufs [][]byte) (found [
 			missing = append(missing, chunks[i])
 			i++
 		} else if chunkKey > keys[j] {
-			level.Debug(util.Logger).Log("msg", "got chunk from cache we didn't ask for")
+			level.Error(util.Logger).Log("msg", "got chunk from cache we didn't ask for")
```
Comment (csmarchbanks) [inline pkg/chunk/chunk_store_utils.go]: Perhaps Info or Warn? This isn't going to cause a request to fail or anything like that.

File: pkg/chunk/cache/cache_test.go [hunk 2]
```diff
@@ -82,12 +83,37 @@ func testCacheMultiple(t *testing.T, cache cache.Cache, keys []string, chunks []
...
+		c, err := chunk.ParseExternalKey(userID, found[i])
+		require.NoError(t, err)
+		err = c.Decode(chunk.NewDecodeContext(), bufs[i])
+		require.NoError(t, err)
+		result = append(result, c)
+	}
+	require.Equal(t, chunks, result)
+}
+
+func testChunkFetcher(t *testing.T, c cache.Cache, keys []string, chunks []chunk.Chunk) {
```
Comment (csmarchbanks) [inline pkg/chunk/cache/cache_test.go]: The fact that the tests for ChunkFetcher live in a different package from the implementation is a bit weird to me. Perhaps some of these tests should live next to the fetcher? WDYT?

The behavior existed before this PR too, so I don't think it necessarily needs to be changed now.

Comment (tomwilkie) [reply pkg/chunk/cache/cache_test.go]: Yeah the challenge is circular imports; I've got a TODO item to complete reorganise the chunk package, as its a bit of a pain.

**Comment by tomwilkie** [comment]
> LGTM, have you measured any performance characteristics yet?

Not really; we saw a gap in our traces for some queries, and that gap is gone, but hard to get concrete results for that.

**Comment by bboreham** [comment]
I think it's worth noting that previously the parallelism of chunk decoding, per process, was "however many queries you are running in parallel", and now it's 16. Which I think is better, but it's not necessarily more than before.

**Pull request merged**